File: app/models/view_override.rb
class ViewOverride < ActiveRecord::Base
  belongs_to :theme
  after_save :initiate

  def initiate
    Deface::Override.new( :virtual_path => self.virtual_path,
                          :name => self.name,
                          self.target.to_sym => self.selector,
                          :closing_selector => self.closing_selector,
                          self.replace_with.to_sym => self.replacement,
                          :disabled => self.disabled,
                          :sequence =>  (self.sequence_target.blank? ? 100 : {self.sequence.to_sym => self.sequence_target}))
  end
end

File: app/assets/javascripts/deface/models/view_override.js
var ViewOverride = Backbone.Model.extend({
  toJSON: function() {
    var object = new Object;
    object['view_override'] = _.clone(this.attributes);
    return object;
  },

  klass: function(){
    return 'view_override'
  },

  defaults: {
    virtual_path: '', 
    target: 'replace', 
    disabled: false, 
    replace_with: 'text', 
    selector: '',
    sequence: 'before', 
    sequence_target: ''
  },

  url : function() {
    var base = '/deface/themes/' + Deface.theme_id + '/view_overrides';
    if (this.isNew()) return base;
    return base + (base.charAt(base.length - 1) == '/' ? '' : '/') + this.id;
  },

  validate: function(attrs) {
    var errors = new Object;
    var has_errors = false;

    if(attrs.name==""){
      errors['name'] = "Name cannot be blank";
      has_errors = true;
    }

    if(attrs.virtual_path==""){
      errors['virtual_path'] = "Virutal Path cannot be blank";
      has_errors = true;
    }

    if(attrs.target==""){
      errors['target'] = "Action cannot be blank";
      has_errors = true;
    }

    if(attrs.selector==""){
      errors['selector'] = "Selector cannot be blank";
      has_errors = true;
    }

    if(has_errors){
      return errors;
    }
  }

});
